crates/core_simd/examples/dot_product.rs

```diff
@@ -0,0 +1,31 @@
+// Code taken from the `packed_simd` crate
+// Run this code with `cargo test --example dot_product`
+#![feature(array_chunks)]
+use core_simd::*;
+
+pub fn dot_prod(a: &[f32], b: &[f32]) -> f32 {
+    assert_eq!(a.len(), b.len());
+
+    // TODO handle remainder when a.len() % 4 != 0
+    a.array_chunks::<4>()
+        .map(|&a| f32x4::from_array(a))
+        .zip(b.array_chunks::<4>().map(|&b| f32x4::from_array(b)))
+        .map(|(a, b)| (a * b).horizontal_sum())
+        .sum()
```

Comment:
Horizontal sums are usually slow so this is probably a better way to do a dot product:

```rust
let mut sum = a.array_chunks::<4>()
    .map(|&a| f32x4::from_array(a))
    .zip(b.array_chunks::<4>().map(|&b| f32x4::from_array(b)))
    .map(|(a, b) a * b)
    .fold(f32x4::splat(0.), std::ops::Add)
    .horizontal_sum();
let remain = a.len() - (a.len() % 4);
sum += a[remain..].zip(&b[remain..]).map((a, b) a * b).sum();
```

The `fold` can be replaced with `sum` once that trait is implemented.

Reply:
Oh that's neat - I think it would be good to document both cases and show performance differences.

Reply:
I might be out of place here, but this seems to allocate more memory than necessary:

`.map(|(a, b) a * b)` will need ~len(a) many allocations, such that the results can be used in the subsequent fold.

Instead you could avoid that by using the `fold` directly:
```rust
let mut sum = a.array_chunks::<4>()
    .map(|&a| f32x4::from_array(a))
    .zip(b.array_chunks::<4>().map(|&b| f32x4::from_array(b)))
    .fold(f32x4::splat(0.), |acc (a, b)| acc + a * b)
    .horizontal_sum();
```
which allocates just 1 XMM register for accumulation (f32x4) and and the op inside the fold can be replaced by a `mul_add` (or some other fused multiply-add) as mentioned in the todo below.

Reply:
I've added both of these versions to the recent PR. Marking as resolved.